File: src/main/java/org/ozwillo/dcimporter/service/PublikService.java
package org.ozwillo.dcimporter.service;

import org.oasis_eu.spring.datacore.DatacoreClient;
import org.oasis_eu.spring.datacore.model.DCOperator;
import org.oasis_eu.spring.datacore.model.DCQueryParameters;
import org.oasis_eu.spring.datacore.model.DCResource;
import org.ozwillo.dcimporter.config.Prop;
import org.ozwillo.dcimporter.model.FormModel;
import org.ozwillo.dcimporter.model.ListFormsModel;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.io.UnsupportedEncodingException;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;
import java.net.URLEncoder;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.*;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import javax.xml.bind.DatatypeConverter;

import org.apache.commons.codec.binary.Base64;

@Service
public class PublikService {

	private static final Logger LOGGER = LoggerFactory.getLogger(PublikService.class);

	private RestTemplate restTemplate = new RestTemplate();

	private final DatacoreClient datacoreClient;
	private final SystemUserService systemUserService;
	private final Prop props;

	@Value("${publik.formTypeEM}")
	private String formTypeEM;
	@Value("${publik.formTypeSVE}")
	private String formTypeSVE;
	@Value("${publik.algo}")
	private String algo;
	@Value("${publik.orig}")
	private String orig;
	@Value("${publik.secret}")
	private String secret;
	@Value("${publik.datacore.project}")
	private String datacoreProject;
	@Value("${publik.datacore.modelEM}")
	private String datacoreModelEM;
	@Value("${publik.datacore.modelSVE}")
	private String datacoreModelSVE;
	@Value("${publik.datacore.modelORG}")
	private String datacoreModelORG;
	@Value("${publik.datacore.modelUser}")
	private String datacoreModelUser;
	@Value("${datacore.baseUri}")
	private String datacoreBaseUri;

	@Autowired
	public PublikService(DatacoreClient datacoreClient, SystemUserService systemUserService, Prop props) {
		this.datacoreClient = datacoreClient;
		this.systemUserService = systemUserService;
		this.props = props;
	}

	private FormModel getForm(String url) throws URISyntaxException {

		URI url_finale = sign_url(url);
		LOGGER.error("URL get Form {}", url_finale);

		return restTemplate.getForObject(url_finale, FormModel.class);
	}

	public void syncPublikForms(String baseUrl, DCResource dcOrganization, String formType) {

		String initUrl = baseUrl + "/api/forms/" + formType + "/list?anonymise";

		try {
			URI url = sign_url(initUrl);
			LOGGER.debug("Calling Publik at URL {}", url);
			ListFormsModel[] forms = restTemplate.getForObject(url, ListFormsModel[].class);

			for (ListFormsModel f : forms) {
				FormModel formModel = getForm(formatUrl(f.getUrl()));
				DCResource dcResource = convertToDCResource(dcOrganization, formModel);
				systemUserService.runAs(() -> datacoreClient.saveResource(datacoreProject, dcResource));
			}
		} catch (URISyntaxException e) {
			LOGGER.error("Exception Uri Syntax : " + e);
		} catch (MalformedURLException e) {
			LOGGER.error("MalformedURLException : " + e);
		}
	}

	public void saveResourceToDC(FormModel form) {

		String orgLegalName = null;
		for (Map<String, String> instance : props.getInstance()) {
			if (form.getUrl().contains(instance.get("baseUrl")))
				orgLegalName = instance.get("organization");
		}

        Optional<DCResource> dcOrganization = getDCOrganization(orgLegalName);
		if (!dcOrganization.isPresent()) {
            LOGGER.error("Unable to get organization {}", orgLegalName);
            return;
        }

		systemUserService.runAs(() -> datacoreClient.saveResource(datacoreProject, convertToDCResource(dcOrganization.get(), form)));
	}

	private DCResource convertToDCResource(DCResource dcOrganization, FormModel form) {

        DCResource dcResource = new DCResource();

		dcResource.setBaseUri(datacoreBaseUri);

		dcResource.setIri(dcOrganization.getIri() + "/" + form.getDisplay_id());

		dcResource.set("citizenreq:displayId", form.getDisplay_id());
		dcResource.set("citizenreq:lastUpdateTime", form.getLast_update_time());
		dcResource.set("citizenreq:displayName", form.getDisplay_name());
		dcResource.set("citizenreq:submissionChannel", form.getSubmission().getChannel());
		dcResource.set("citizenreq:submissionBackoffice", form.getSubmission().getBackoffice().toString());
		dcResource.set("citizenreq:url", form.getUrl());
		dcResource.set("citizenreq:receiptTime", form.getReceipt_time());

		dcResource.set("citizenreq:criticalityLevel", form.getCriticality_level().toString());
		dcResource.set("citizenreq:id", form.getId());
		dcResource.set("citizenreq:organization", dcOrganization.getUri());

		if (form.getUser() != null)
			dcResource.set("citizenreq:user", createUserDCResource(form));

		if (isEMrequest(form)) {
			convertToDCResourceEM(dcResource, form);
		} else if (isSVErequest(form)) {
			convertToDCResourceSVE(dcResource, form);
		}
		LOGGER.debug("DCResouce --> :" + dcResource.toString());
		return dcResource;
	}

	private DCResource convertToDCResourceEM(DCResource dcResource, FormModel form) {

		dcResource.setType(datacoreModelEM);

		dcResource.set("citizenreqem:familyName", form.getFields().get("nom_famille").toString());
		dcResource.set("citizenreqem:firstName", form.getFields().get("prenom").toString());
		dcResource.set("citizenreqem:phone", form.getFields().get("telephone").toString());
		if (form.getFields().get("detail") != null)
			dcResource.set("citizenreqem:detail", form.getFields().get("detail").toString());
		dcResource.set("citizenreqem:objectSummary", form.getFields().get("objet_rendez_vous_raw").toString());
		dcResource.set("citizenreqem:objectDetail", form.getFields().get("objet_rendez_vous").toString());
		dcResource.set("citizenreqem:desiredDate", form.getFields().get("date_souhaitee").toString());
		dcResource.set("citizenreqem:email", form.getFields().get("courriel").toString());

		return dcResource;
	}

	private DCResource convertToDCResourceSVE(DCResource dcResource, FormModel form) {

		dcResource.setType(datacoreModelSVE);

		dcResource.set("citizenreqsve:title", form.getFields().get("civilite").toString());
		dcResource.set("citizenreqsve:familyName", form.getFields().get("nom").toString());
		dcResource.set("citizenreqsve:firstName", form.getFields().get("prenoms").toString());
		dcResource.set("citizenreqsve:email", form.getFields().get("email").toString());
		dcResource.set("citizenreqsve:streetAddress", form.getFields().get("voie").toString());
		dcResource.set("citizenreqsve:zipCode", form.getFields().get("code_postal").toString());
		dcResource.set("citizenreqsve:city", form.getFields().get("commune").toString());
		dcResource.set("citizenreqsve:entityType", form.getFields().get("entite_raw").toString());
		dcResource.set("citizenreqsve:object", form.getFields().get("objet").toString());
		dcResource.set("citizenreqsve:message", form.getFields().get("message").toString());

		@SuppressWarnings("unchecked")
		HashMap<String, String> doc = (HashMap<String, String>) form.getFields().get("doc");
		if (doc != null) {
			dcResource.set("citizenreqsve:docContent", doc.get("content"));
			dcResource.set("citizenreqsve:docFieldId", doc.get("field_id"));
			dcResource.set("citizenreqsve:docContentType", doc.get("content_type"));
			dcResource.set("citizenreqsve:docFileName", doc.get("filename"));
		}
		if (form.getFields().get("siret") != null)
			dcResource.set("citizenreqsve:siret", form.getFields().get("siret").toString());
		if (form.getFields().get("siret_entreprise") != null)
			dcResource.set("citizenreqsve:siret", form.getFields().get("siret_entreprise").toString());
		if (form.getFields().get("rna") != null)
			dcResource.set("citizenreqsve:rna", form.getFields().get("rna").toString());
		if (form.getFields().get("nom_entite_entreprise") != null)
			dcResource.set("citizenreqsve:legalName", form.getFields().get("nom_entite_entreprise").toString());
		if (form.getFields().get("nom_entite") != null)
			dcResource.set("citizenreqsve:entityName", form.getFields().get("nom_entite").toString());
		return dcResource;
	}

	private String createUserDCResource(FormModel form) {

		DCResource dcResource = new DCResource();

		dcResource.setBaseUri(datacoreBaseUri);
		dcResource.setType(datacoreModelUser);
		dcResource.setIri(form.getUser().getNameID()[0]);

		systemUserService.runAs(() -> {
			if (datacoreClient.getResourceFromURI(datacoreProject, dcResource.getUri()).getResource() == null) {

				dcResource.set("citizenrequser:email", form.getUser().getEmail());
				dcResource.set("citizenrequser:nameID", form.getUser().getNameID()[0]);
				dcResource.set("citizenrequser:userId", form.getUser().getId().toString());
				dcResource.set("citizenrequser:name", form.getUser().getName());
				datacoreClient.saveResource(datacoreProject, dcResource);
			}
		});

		return dcResource.getUri();
	}

	public Optional<DCResource> getDCOrganization(String orgLegalName) {
        DCQueryParameters queryParametersOrg = new DCQueryParameters("org:legalName", DCOperator.EQ, orgLegalName);
        List<DCResource> resources = new ArrayList<>();
		systemUserService.runAs(() -> {
			resources.addAll(datacoreClient
					.findResources(datacoreProject, datacoreModelORG, queryParametersOrg, 0, 1));
		});

		return resources.isEmpty() ? Optional.empty() : Optional.of(resources.get(0));
    }

	/**
	 * Calculate a signature with sha256
	 * 
	 * @return
	 */
	private String calculateSignature(String message, String key) {

		try {

			Mac sha256_HMAC = Mac.getInstance("HmacSHA256");
			SecretKeySpec secret_key = new SecretKeySpec(key.getBytes(), "HmacSHA256");
			sha256_HMAC.init(secret_key);

			String hash = Base64.encodeBase64String(sha256_HMAC.doFinal(message.getBytes()));
			LOGGER.debug("Signature : " + hash);

			hash = URLEncoder.encode(hash, "UTF-8");
			LOGGER.debug("URL encoded hash : " + hash);

			return hash;
		} catch (Exception e) {
			LOGGER.error("Exception when calculate the signature : " + e);
			return null;
		}

	}

	private URI sign_url(String url) throws URISyntaxException {

		TimeZone tz = TimeZone.getTimeZone("UTC");
		DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'");
		df.setTimeZone(tz);
		String thisMoment = df.format(new Date());

		Random random = new Random();
		// create byte array
		byte[] nbyte = new byte[16];
		// put the next byte in the array
		random.nextBytes(nbyte);
		String nonce = DatatypeConverter.printHexBinary(nbyte);

		try {
			String newQuery = "";
			URL parsedUrl = new URL(url);

			if (parsedUrl.getQuery() != null)
				newQuery = parsedUrl.getQuery() + "&";

			newQuery += "algo=" + this.algo + "&timestamp=" + URLEncoder.encode(thisMoment, "UTF-8") + "&nonce=" + nonce
					+ "&orig=" + URLEncoder.encode(this.orig, "UTF-8");
			String signature = calculateSignature(newQuery, this.secret);
			newQuery += "&signature=" + signature;

			return new URI(
					parsedUrl.getProtocol() + "://" + parsedUrl.getHost() + parsedUrl.getPath() + "?" + newQuery);
		} catch (MalformedURLException e) {
			LOGGER.error("MalformedURLException : " + e);
			return null;
		} catch (UnsupportedEncodingException e) {
			LOGGER.error("Unsupported encoding exception !?");
			return null;
		}
	}

	private String formatUrl(String url_base) throws MalformedURLException {

		URL parsedUrl = new URL(url_base);
		String url_finale = "https://" + parsedUrl.getHost() + "/api/forms" + parsedUrl.getPath();
		return url_finale;
	}

	private Boolean isEMrequest(FormModel form) {

		if (form.getId().contains(this.formTypeEM)) {
			return true;
		}
		return false;
	}

	private Boolean isSVErequest(FormModel form) {

		if (form.getId().contains(this.formTypeSVE)) {
			return true;
		}
		return false;
	}

}

File: src/main/kotlin/org/ozwillo/dcimporter/service/PublikService.kt
package org.ozwillo.dcimporter.service

import com.fasterxml.jackson.annotation.JsonProperty
import org.apache.commons.lang3.RandomStringUtils
import org.ozwillo.dcimporter.config.DatacoreProperties
import org.ozwillo.dcimporter.model.BusinessAppConfiguration
import org.ozwillo.dcimporter.model.BusinessMapping
import org.ozwillo.dcimporter.model.datacore.*
import org.ozwillo.dcimporter.model.publik.FormModel
import org.ozwillo.dcimporter.model.publik.ListFormsModel
import org.ozwillo.dcimporter.model.publik.User
import org.ozwillo.dcimporter.repository.BusinessAppConfigurationRepository
import org.ozwillo.dcimporter.repository.BusinessMappingRepository
import org.ozwillo.dcimporter.util.hmac
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.http.RequestEntity
import org.springframework.stereotype.Service
import org.springframework.util.LinkedMultiValueMap
import org.springframework.web.client.HttpClientErrorException
import org.springframework.web.client.RestTemplate
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.util.DefaultUriBuilderFactory
import reactor.core.publisher.Mono
import java.net.MalformedURLException
import java.net.URI
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*

// TODO : needs a lot of cleanup and refactoring !!!!
@Service
class PublikService(
    private val datacoreService: DatacoreService,
    private val businessAppConfigurationRepository: BusinessAppConfigurationRepository,
    private val businessMappingRepository: BusinessMappingRepository,
    private val datacoreProperties: DatacoreProperties
) {

    @Value("\${publik.formTypeEM}")
    private val formTypeEM: String? = null
    @Value("\${publik.formTypeSVE}")
    private val formTypeSVE: String? = null
    @Value("\${publik.algo}")
    private val algo: String? = null
    @Value("\${publik.orig}")
    private val orig: String = "ozwillo-dcimporter"
    @Value("\${datacore.model.project}")
    private val datacoreProject: String = "datacoreProject"
    @Value("\${datacore.model.modelEM}")
    private val datacoreModelEM: String? = null
    @Value("\${datacore.model.modelSVE}")
    private val datacoreModelSVE: String? = null
    @Value("\${datacore.model.modelUser}")
    private val datacoreModelUser: String = "citizenreq:user_0"
    @Value("\${datacore.model.modelORG}")
    private val datacoreModelORG: String = "org:Organization_0"

    private fun getForm(url: String, secret: String): Mono<FormModel> {

        val signedQuery = signQuery("email=admin@ozwillo-dev.eu&", secret)
        LOGGER.debug("Getting Publik form at URL $url?$signedQuery")

        val uriBuilderFactory = DefaultUriBuilderFactory()
        uriBuilderFactory.encodingMode = DefaultUriBuilderFactory.EncodingMode.NONE
        val clientV2 = WebClient.builder()
            .uriBuilderFactory(uriBuilderFactory)
            .build()
        return clientV2.get()
            .uri("$url?$signedQuery")
            .retrieve()
            .onStatus(HttpStatus::is4xxClientError,
                { response ->
                    response.bodyToMono(PublikResponse::class.java).map {
                        RuntimeException(it.toString())
                    }
                })
            .bodyToMono(FormModel::class.java)
    }

    fun syncPublikForms(
        businessAppConfiguration: BusinessAppConfiguration,
        dcOrganization: DCResource,
        formType: String
    ): Mono<HttpStatus> {

        val signedQuery = signQuery("email=admin@ozwillo-dev.eu&", businessAppConfiguration.secretOrToken!!)

        LOGGER.debug(
            "Getting Publik form list at URL ${businessAppConfiguration.baseUrl}/api/forms/$formType/list?$signedQuery")

        val uriBuilderFactory = DefaultUriBuilderFactory()
        uriBuilderFactory.encodingMode = DefaultUriBuilderFactory.EncodingMode.NONE
        val clientV2 = WebClient.builder()
            .uriBuilderFactory(uriBuilderFactory)
            .build()
        return clientV2.get()
            .uri("${businessAppConfiguration.baseUrl}/api/forms/$formType/list?$signedQuery")
            .retrieve()
            .onStatus(HttpStatus::is4xxClientError) { response ->
                response.bodyToMono(PublikResponse::class.java).map {
                    RuntimeException(it.toString())
                }
            }
            .bodyToFlux(ListFormsModel::class.java)
            .flatMap { listFormsModel ->
                getForm(
                    formatUrl(listFormsModel.url),
                    businessAppConfiguration.secretOrToken
                )
            }
            .flatMap { formModel -> convertToDCResource(dcOrganization, formModel) }
            .map { datacoreService.saveResource(datacoreProject, it.first, it.second, null) }
            .count()
            .map { HttpStatus.OK }
    }

    data class PublikResponse(
        @JsonProperty("err_class") val errClass: String,
        @JsonProperty("err_desc") val errDesc: String,
        @JsonProperty("err") val err: Int
    )

    fun formToDCResource(organizationSiret: String, form: FormModel): Mono<Pair<DCModelType, DCResource>> {

        LOGGER.debug("Got form from Publik : $form")

        return datacoreService.getResourceFromIRI(datacoreProject, datacoreModelORG, "FR/$organizationSiret", null)
            .flatMap { convertToDCResource(it, form) }
            .map { result ->
                val businessMapping = BusinessMapping(
                    applicationName = name, businessId = form.url,
                    dcId = result.second.getUri(), type = result.first
                )
                businessMappingRepository.save(businessMapping).subscribe()

                result
        }
    }

    private fun convertToDCResource(dcOrganization: DCResource, form: FormModel): Mono<Pair<DCModelType, DCResource>> {

        return getOrCreateUser(form).map { userUri ->
            val type = if (isEMrequest(form)) datacoreModelEM else datacoreModelSVE

            val dcFormResource = DCResource(datacoreProperties.baseResourceUri(),
                type = type!!, iri = dcOrganization.getIri() + "/" + form.display_id)

            dcFormResource.setStringValue("citizenreq:displayId", form.display_id)
            dcFormResource.setStringValue("citizenreq:lastUpdateTime", form.last_update_time)
            dcFormResource.setStringValue("citizenreq:displayName", form.display_name)
            dcFormResource.setStringValue("citizenreq:submissionChannel", form.submission.channel)
            dcFormResource.setStringValue("citizenreq:submissionBackoffice", form.submission.backoffice.toString())
            dcFormResource.setStringValue("citizenreq:url", form.url)
            dcFormResource.setStringValue("citizenreq:receiptTime", form.receipt_time)
            dcFormResource.setStringValue("citizenreq:workflowStatus", form.workflowStatus.orEmpty())

            dcFormResource.setStringValue("citizenreq:criticalityLevel", form.criticality_level.toString())
            dcFormResource.setStringValue("citizenreq:id", form.id)
            dcFormResource.setStringValue("citizenreq:organization", dcOrganization.getUri())

            // TODO shouldn't user be instead the guy who issued the request ?
            // Currently it is the agent who triggered the status change
            dcFormResource.setStringValue("citizenreq:user", userUri)

            if (isEMrequest(form)) {
                convertToDCResourceEM(dcFormResource, form)
            } else if (isSVErequest(form)) {
                convertToDCResourceSVE(dcFormResource, form)
            }

            Pair(type, dcFormResource)
        }

    }

    private fun convertToDCResourceEM(dcResource: DCResource, form: FormModel): DCResource {

        dcResource.setStringValue("citizenreqem:familyName", form.fields["nom"].toString())
        dcResource.setStringValue("citizenreqem:firstName", form.fields["prenom"].toString())
        dcResource.setStringValue("citizenreqem:phone", form.fields["telephone"].toString())
        dcResource.setStringValue("citizenreqem:email", form.fields["email"].toString())
        if (form.fields["detail"] != null)
            dcResource.setStringValue("citizenreqem:detail", form.fields["detail"].toString())
        dcResource.setStringValue("citizenreqem:objectSummary", form.fields["objet_raw"].toString())
        dcResource.setStringValue("citizenreqem:objectDetail", form.fields["objet"].toString())
        if (form.fields["date_souhaitee"] != null)
            dcResource.setStringValue("citizenreqem:desiredDate", form.fields["date_souhaitee"].toString())

        if (form.fields["courrier"] != null) {
            val courrierFieds: Map<String, Any> = form.fields["courrier"] as Map<String, Any>
            dcResource.setStringValue("citizenreqem:fileContentType", courrierFieds["content_type"].toString())
            dcResource.setStringValue("citizenreqem:fileName", courrierFieds["filename"].toString())
            dcResource.setStringValue("citizenreqem:fileContent", courrierFieds["content"].toString())
        }

        return dcResource
    }

    private fun convertToDCResourceSVE(dcResource: DCResource, form: FormModel): DCResource {

        dcResource.setStringValue("citizenreqsve:title", form.fields["civilite"].toString())
        dcResource.setStringValue("citizenreqsve:familyName", form.fields["nom"].toString())
        dcResource.setStringValue("citizenreqsve:firstName", form.fields["prenoms"].toString())
        dcResource.setStringValue("citizenreqsve:email", form.fields["email"].toString())
        dcResource.setStringValue("citizenreqsve:streetAddress", form.fields["voie"].toString())
        dcResource.setStringValue("citizenreqsve:zipCode", form.fields["code_postal"].toString())
        dcResource.setStringValue("citizenreqsve:city", form.fields["commune"].toString())
        dcResource.setStringValue("citizenreqsve:entityType", form.fields["entite_raw"].toString())
        dcResource.setStringValue("citizenreqsve:object", form.fields["objet"].toString())
        dcResource.setStringValue("citizenreqsve:message", form.fields["message"].toString())

        val doc = form.fields["doc"] as HashMap<String, String>
        dcResource.setStringValue("citizenreqsve:docContent", doc["content"].orEmpty())
        dcResource.setStringValue("citizenreqsve:docFieldId", doc["field_id"].orEmpty())
        dcResource.setStringValue("citizenreqsve:docContentType", doc["content_type"].orEmpty())
        dcResource.setStringValue("citizenreqsve:docFileName", doc["filename"].orEmpty())

        if (form.fields["siret"] != null)
            dcResource.setStringValue("citizenreqsve:siret", form.fields["siret"].toString())
        if (form.fields["siret_entreprise"] != null)
            dcResource.setStringValue("citizenreqsve:siret", form.fields["siret_entreprise"].toString())
        if (form.fields["rna"] != null)
            dcResource.setStringValue("citizenreqsve:rna", form.fields["rna"].toString())
        if (form.fields["nom_entite_entreprise"] != null)
            dcResource.setStringValue("citizenreqsve:legalName", form.fields["nom_entite_entreprise"].toString())
        if (form.fields["nom_entite"] != null)
            dcResource.setStringValue("citizenreqsve:entityName", form.fields["nom_entite"].toString())
        return dcResource
    }

    fun getOrCreateUser(form: FormModel): Mono<String> {

        // for now, let's say agent_sictiam is the universal fallback user
        val nameId = if (form.user == null) "5c977a7f1d444fa1ab0f777325fdda93" else form.user.nameID[0]

        return datacoreService.getResourceFromIRI(datacoreProject, datacoreModelUser, nameId, null)
            .map { it.getUri() }
            .switchIfEmpty(Mono.just(1).flatMap { createUser(form.user!!) })
    }

    private fun createUser(user: User): Mono<String> {
        val dcUserResource = DCResource(datacoreProperties.baseResourceUri(),
            type = datacoreModelUser, iri = user.nameID[0])
        dcUserResource.setStringValue("citizenrequser:email", user.email)
        dcUserResource.setStringValue("citizenrequser:nameID", user.nameID[0])
        dcUserResource.setStringValue("citizenrequser:userId", user.id.toString())
        dcUserResource.setStringValue("citizenrequser:name", user.name)
        return datacoreService.saveResource(datacoreProject, datacoreModelUser, dcUserResource, null)
            .map { it.getUri() }
    }

    data class PublikStatusResponse(val url: String?, val err: Int?)

    fun changeStatus(siret: String, dcResource: DCResource) {

        if (dcResource.getStringValue("citizenreq:workflowStatus") != "Terminé") {
            LOGGER.debug("Ignoring unhandled workflow status ${dcResource.getStringValue("citizenreq:workflowStatus")}")
            return
        }

        val businessAppConfigurationMono: Mono<BusinessAppConfiguration> =
            businessAppConfigurationRepository.findByOrganizationSiretAndApplicationName(siret, name)
        val businessMappingMono: Mono<BusinessMapping> =
            businessMappingRepository.findByDcIdAndApplicationName(dcResource.getUri(), name)

        businessAppConfigurationMono
            .zipWith(businessMappingMono)
            .subscribe { tuple2 ->
                val businessAppConfiguration = tuple2.t1
                val businessMapping = tuple2.t2

                val signedQuery =
                    signQuery("email=${businessAppConfiguration.login}&", businessAppConfiguration.secretOrToken!!)
                val uri = "${businessMapping.businessId}jump/trigger/close?$signedQuery"

                val restTemplate = RestTemplate()
                val headers = LinkedMultiValueMap<String, String>()
                headers.set("Accept", "application/json")
                val request = RequestEntity<Any>(null, headers, HttpMethod.POST, URI(uri))

                try {
                    val response = restTemplate.exchange(request, PublikStatusResponse::class.java)
                    LOGGER.debug("Got Publik response for status change $response")
                } catch (e: HttpClientErrorException) {
                    // TODO : what can we do here ?
                    LOGGER.warn("Request status change failed in Publik ${e.responseBodyAsString}")
                }
            }
    }

    private fun signQuery(query: String, secret: String): String {

        val tz = TimeZone.getTimeZone("UTC")
        val df = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'").apply { timeZone = tz }
        val thisMoment = df.format(Date())

        val nonce = RandomStringUtils.random(64, true, true)

        val fullQuery = query + "algo=" + this.algo +
                "&timestamp=" + thisMoment +
                "&nonce=" + nonce +
                "&orig=" + this.orig
        val signature = fullQuery.hmac("HmacSHA256", secret)
        return "$fullQuery&signature=$signature"
    }

    @Throws(MalformedURLException::class)
    private fun formatUrl(url_base: String): String {

        LOGGER.debug("Formatting URL $url_base")
        val parsedUrl = URL(url_base)
        return "https://" + parsedUrl.host + "/api/forms" + parsedUrl.path
    }

    private fun isEMrequest(form: FormModel): Boolean {
        LOGGER.debug("Guessing type from ${form.url}")
        return form.url.contains(this.formTypeEM!!)
    }

    private fun isSVErequest(form: FormModel): Boolean {
        return form.id.contains(this.formTypeSVE!!)
    }

    companion object {

        private val LOGGER = LoggerFactory.getLogger(PublikService::class.java)
        const val name: String = "publik"
    }

}
